### DISCUSSION TITLE
It is possible to put parentheses in WHERE condition without CustomExpression

### DISCUSSION THREAD

**Discussion by ycherniavskyi** [simolus3/drift#1544] [Q&A] [answered]
Particularly is it possible to write the next WHERE condition with dart code and without `CustomExpression`:
```dart
  SimpleSelectStatement<$ContactsTableTable, ContactData> _selectAllContacts([ContactSourceType? sourceType]) =>
      select(contactsTable)
        ..where((t) {
          if (sourceType == null) {
            return const Constant(true);
          } else {
            return t.sourceType.equals(sourceType.index);
          }
        })
        ..where((t) => const CustomExpression<bool?>(
            "(display_name IS NOT NULL AND NOT (display_name = '')) OR (first_name IS NOT NULL AND NOT (first_name = '')) OR (last_name IS NOT NULL AND NOT (last_name = ''))"))
        ..orderBy([
          (t) => OrderingTerm.asc(t.displayName),
          (t) => OrderingTerm.asc(t.lastName),
          (t) => OrderingTerm.asc(t.firstName),
        ]);

  Stream<List<ContactData>> watchAllContacts([ContactSourceType? sourceType]) => _selectAllContacts(sourceType).watch();
```

**Comment by simolus3** [comment, marked as answer]
Drift is supposed to add parentheses where this is necessary. However, `OR`  has the lowest precedence and `NOT` has a lower precedence than `=`. So it looks like the parentheses are not needed in your custom expression? Do you just want the parentheses to make the generated query easier to reason about or is there something going wrong here?

**Reply by ycherniavskyi** [reply]
Wow, just recheck operators precedences of my query and yes it could be rewritten without parentheses 😊.
```dart
        ..where((t) =>
            t.displayName.isNotNull() & t.displayName.equalsExp(const Constant('')).not() |
            t.firstName.isNotNull() & t.firstName.equalsExp(const Constant('')).not() |
            t.lastName.isNotNull() & t.lastName.equalsExp(const Constant('')).not());
```

But what if for example, I will need something like this: `(display_name = 'A' OR display_name = 'B') AND (first_name = 'C' OR first_name = '')`. Can I write It with one `where` function?

**Reply by simolus3** [reply]
> Can I write It with one `where` function?

You can :D I added a test to be sure in 05a8b9abf1952b22d7cbeb8d56a88b207968710f.

**Reply by ycherniavskyi** [reply]
Wow. I try to use parentheses in code with my initial condition, and because they were absent in the final query, I think it is impossible. But now I understand that my parentheses were removed because of redundancy (according to operators precedence).
Perfect!
